src/Microsoft.ML.LightGBM/WrappedLightGbmDataset.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading.Tasks;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.Runtime.LightGBM
+{
+    /// <summary>
+    /// Wrapper of Dataset object of LightGBM.
+    /// </summary>
+    internal sealed class Dataset : IDisposable
+    {
+        private IntPtr _handle;
+        private int _lastPushedRowID;
+        public IntPtr Handle => _handle;
+
+        public unsafe Dataset(double[][] sampleValuePerColumn,
+            int[][] sampleIndicesPerColumn,
+            int numCol,
+            int[] sampleNonZeroCntPerColumn,
+            int numSampleRow,
+            int numTotalRow,
+            string param, float[] labels, float[] weights = null, int[] groups = null)
+        {
+            _handle = IntPtr.Zero;
+
+            // Use GCHandle to pin the memory, avoid the memory relocation.
+            GCHandle[] gcValues = new GCHandle[numCol];
+            GCHandle[] gcIndices = new GCHandle[numCol];
+            try
+            {
+                double*[] ptrArrayValues = new double*[numCol];
+                int*[] ptrArrayIndices = new int*[numCol];
+                Parallel.For(0, numCol, i =>
```

Comment:
There appears to be a few misuses of `Parallel.For` in this code.

Usually if the operation inside the loop is very small, the overhead of doing the Thread management and communication outweighs any parallel savings. In this case, it is over 2x slower using `Parallel.For` than just a simple for loop.  I write a simple benchmark of the two: https://gist.github.com/eerhardt/99071a3b6b00ddfe5ae02d9f85d4f9d6.

Results:

```
BenchmarkDotNet=v0.10.14, OS=Windows 10.0.17134
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.300
  [Host]     : .NET Core 2.1.0 (CoreCLR 4.6.26515.07, CoreFX 4.6.26515.06), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0 (CoreCLR 4.6.26515.07, CoreFX 4.6.26515.06), 64bit RyuJIT
```

|     Method |     Mean |     Error |   StdDev |
|----------- |---------:|----------:|---------:|
|   Original | 62.12 us | 1.2207 us | 1.199 us |
| NoParallel | 27.06 us | 0.5554 us | 1.196 us |

Reply:
This is great! Thank you @eric.erhardt@microsoft.com for going an extra mile to validate parallel for loop.

---
In reply to: [197847135](https://github.com/dotnet/machinelearning/pull/392#discussion_r197847135) [](ancestors = 197847135)